### PR TITLE
c++ version segmenter

### PR DESCRIPTION
**Pull request by LvHang**
After fixing some bugs and doing some tests, I think this is a right version.
If you encounter any problem when you use it, please let me know. Thanks.

### CODE REVIEW

**Comment by hhadian** [comment]
Thanks for the PR Hang! I'll merge and test it tomorrow..

**Comment by hhadian** [comment]
Hang, just wanted to let you know that I am working on this but there is a problem: to be able to run c-segmenter, I had to switch to python3 (as you suggested) but that has caused other issues. I am re-training the models to see if the problems disappear...

**Comment by danpovey** [comment]
Is python3 supposed to be a requirement to use the C++ segmenter?  I.e. is
there a known reason for the dependency?

**Comment by LvHang** [comment]
Actually, I use "cython"(http://cython.org/) as a bridge to connect the python code and C++ code.
I guess "cython" version may brings the limitation.
When I use python2.7 to compile the code, it has the following problem --"undefined symbol: _Py_ZeroStruct". But when I use python3, it is ok.

**Comment by danpovey** [comment]
From this:
https://stackoverflow.com/questions/44737063/cython-bbox-so-undefined-symbol-py-zerostruct

"This error message is typical when there's a version mismatch in Python
modules. It could something like cython_bbox.so was compiled and linked
against Python 2.7 and CNTK was compiled against Python 3.5. It's hard to
say exactly what the reason is but you can try to do make inside a python
environment that matches your CNTK version and see if that works"

Is this on CLSP, and how are you invoking cython?

In the package list `dpkg -l | grep cython` it looks like the cython
version installed at Hopkins may be for 2.7 because there is another one
called cython3.

root@b12:~# apt-cache search cython  | grep 'C-Ext'
cython - C-Extensions for Python
cython-dbg - C-Extensions for Python - debug build
cython-doc - C-Extensions for Python - documentation
cython3 - C-Extensions for Python 3
cython3-dbg - C-Extensions for Python 3 - debug build

# show installed version
root@b12:~# dpkg -l |grep cython
ii  cython                                    0.25.2-1
          amd64        C-Extensions for Python

**Comment by LvHang** [comment]
I also saw this stackoverflow. But I'm not sure what causes the version mismatch.

The command to compile the code is-- python setup.py build_ext --inplace.
If I set my python path as: 
PYTHONPATH=/export/a11/hlyu/new/tools/tensorflow_build/.local/lib/python2.7/site-packages/:$PYTHONPATH
export PYTHONPATH=/home/hlyu/.local/lib/python2.7/site-packages:$PYTHONPATH:$PATH
If I use this path, I will encounter that problem.

If I set my python path as:
export PATH="/home/yshao/miniconda3/bin:$PATH"                                  
export PYTHONPATH="${PYTHONPATH}:/home/yshao/miniconda3:$HOME/.local/lib/python3.5/site-packages"
If I use this path, it will be OK. 

@hhadian 
Could you give me a model which you trained with python2.7. 
Yiwen only has the python3 model. I want to use it to figure out the reason.

**Comment by hhadian** [comment]
Sorry, I don't have python2.7 models. I imagined that I had trained using python2 but I had not.

Actually my models were trained using python3 but when compiling the `csegmenter`, cython2.7 was invoked (by python), even though I had set the path to use miniconda3 (which is python3). Not sure what happened.
Finally, I compiled it by explicitly invoking python3.

Anyway, I did a few tests on Madcat Arabic using the csegmenter which was very fast. The objective function numbers and results match the python version's. There are small differences but they might be due to numerical precision. I'll review the code next...

**Comment by LvHang** [comment]
Thanks! I will try to figure out the reason for the dependency.

**Comment by hhadian** [comment]
> Actually my models were trained using python3 but when compiling the csegmenter, cython2.7 was invoked (by python), even though I had set the path to use miniconda3 (which is python3). Not sure what happened

...oh I figured out what happened. It seems that Miniconda3 has both `python` and `python3` in its `bin` directory and `bin/python` actually runs python2.7!
I guess it would make sense to update this line in the README file:
`python setup.py build_ext --inplace` 
to
`python3 setup.py build_ext --inplace` 
because all other scripts in Waldo are by default using python3.

**Comment by danpovey** [comment]
Find-- If the other scripts are for python3, better to standardize on it.
so do as Hossein suggests and invoke as python3.

File: scripts/csegment/segment.h [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    IntPairEqual> pixel2obj;
+  unordered_map<size_t, AdjacencyRecord*> adjacency_records;
+  priority_queue<pair<float, AdjacencyRecord*>, 
+    vector<pair<float, AdjacencyRecord*> >, PriorityCompare> segmenter_queue;
+
+  // AdjacencyRecord is allocated on heap. After merge objects, some
+  // AdjacencyRecord will be only appear in priority queue with minimal
+  // merge_priority value. This vector is used to record all the
+  // AdjacencyRecords, and is used to release them
+  vector<AdjacencyRecord*> adjacency_space;
```
Comment (hhadian) [inline scripts/csegment/segment.h]: I'm not sure if this is needed. How about freeing the adjacency_records when they are being merged (at L568)? I guess that should work.

File: scripts/csegment/segment.cc [hunk 2]
```diff
@@ -0,0 +1,681 @@
+// segment.cc
```
Comment (hhadian) [inline scripts/csegment/segment.cc]: I guess it's better to move the `csegment` directory to `scripts/waldo` and maybe rename it to `csegmenter`.

**Comment by LvHang** [comment]
@hhadian 
I have modified the codes according to your comments. Thanks a lot!

**Comment by danpovey** [comment]
@Hossein Hadian <hhadian@jhu.edu>  please merge when you think it's OK.

**Comment by hhadian** [comment]
Assuming it's tested (with the latest commits), it seems OK to me. 
Thanks Hang, merging...

**Comment by hhadian** [comment]
Something I forgot: I didn't see the `Prune` function  in the C++ version (you can see it in the Python version). Since pruning is used in the Madcat recipe, when/if you have time could you please implement that function too?

**Comment by YiwenShaoStephen** [comment]
@hhadian I think the reason might be that I am still trying new Prune functions since the old one is not able to fix many issues like the mismatch between offset output and class output. I plan to upload this part after getting a satisfactory result on cityscape dataset. But if it is urgent for other projects, I think @LvHang can upload it immediately.

**Comment by LvHang** [comment]
I will add one today.
After Yiwen find a satisfactory plan, we can update it.

**Comment by hhadian** [comment]
No, it's not urgent at all. I just wanted to mention it so we won't forget about it.

**Pull request merged**